### PR TITLE
TYP: Fix type hint for params passed to read_sql_query

### LINKED ISSUE #49716: BUG: pandas.read_sql_query type hint only allows str params
### Pandas version checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.

- [X] I have confirmed this bug exists on the main branch of pandas.

### Reproducible Example

```python
import pandas
import sqlalchemy.engine


if __name__ == '__main__':
    sql_connection: str = "driver={ODBC Driver 17 for SQL Server};server=myServer;database=myDatabase;trusted_connection=yes"
    sql_engine: sqlalchemy.engine = sqlalchemy.create_engine(
        url=sqlalchemy.engine.URL.create(drivername="mssql+pyodbc", query={"odbc_connect": sql_connection})
    )
    with sql_engine.connect() as connection:
        myarg1: int = 10
        myarg2: bool = True
        result: pandas.DataFrame = pandas.read_sql_query(
            sql="""
                DECLARE @arg1 INT = ?;
                DECLARE @arg2 INT = ?
                
                SELECT
                    arg1 = @arg1,
                    arg2 = @arg2
            """,
            con=connection,
            params=[myarg1, myarg2]
        )
        print(result)
```
```


### Issue Description

Pandas 1.4.4 allowed pandas.read_sql_query to be called with params that included int, bool, datetime.date, etc. Pandas 1.5.1 only allows params  that are str. Code that attempts to pass int or bool to a SQL script that is expecting INT parameters fails code inspection. For example, Code Inspection in PyCharm reports the following for the call to pandas.read_sql_query:

Unexpected type(s): (str, Any, list[Union[int, bool]]) Possible type(s): (Any, Any, Union(list[str], dict[str, str], None]) (Any, Any, Union(list[str], dict[str, str], None])

This new behavior is a regression,

### Expected Behavior

The  type hint for params should be defined as typing.Optional[typing.Union[typing.List, typing.Mapping]] so that list of any objects or any dicts can be passed as was t he case with Pandas 1.4.4.


### Installed Versions

<details>

INSTALLED VERSIONS
------------------
commit           : 91111fd99898d9dcaa6bf6bedb662db4108da6e6
python           : 3.9.12.final.0
python-bits      : 64
OS               : Windows
OS-release       : 10
Version          : 10.0.19045
machine          : AMD64
processor        : Intel64 Family 6 Model 78 Stepping 3, GenuineIntel
byteorder        : little
LC_ALL           : None
LANG             : None
LOCALE           : English_United States.1252

pandas           : 1.5.1
numpy            : 1.23.4
pytz             : 2022.6
dateutil         : 2.8.2
setuptools       : 65.5.0
pip              : 22.2.2
Cython           : None
pytest           : None
hypothesis       : None
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : None
html5lib         : None
pymysql          : None
psycopg2         : None
jinja2           : None
IPython          : None
pandas_datareader: None
bs4              : None
bottleneck       : None
brotli           : None
fastparquet      : None
fsspec           : None
gcsfs            : None
matplotlib       : None
numba            : None
numexpr          : None
odfpy            : None
openpyxl         : None
pandas_gbq       : None
pyarrow          : None
pyreadstat       : None
pyxlsb           : None
s3fs             : None
scipy            : None
snappy           : None
sqlalchemy       : 1.4.44
tables           : None
tabulate         : None
xarray           : None
xlrd             : None
xlwt             : None
zstandard        : None
tzdata           : None

</details>

### PR DESCRIPTION
**Pull request by Tolker-KU**
- [x] closes #49716
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [x] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [x] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

### CODE REVIEW

File: pandas/io/sql.py [hunk 1]
```diff
@@ -357,7 +358,7 @@ def read_sql_query(
     con,
     index_col: str | list[str] | None = ...,
     coerce_float=...,
-    params: list[str] | dict[str, str] | None = ...,
+    params: list[Any] | Mapping[str, Any] | None = ...,
```
[the remark this answers is not in the document]
Comment (Tolker-KU) [reply pandas/io/sql.py:361]: I think this is one of the rare cases where `Any` is correct. See the type hint in SQLAlchemy which we are calling into here:
https://github.com/sqlalchemy/sqlalchemy/blob/e79ab08165e01dc7af50fcffadb31468ace51b6c/lib/sqlalchemy/engine/base.py#L1370-L1376
https://github.com/sqlalchemy/sqlalchemy/blob/e79ab08165e01dc7af50fcffadb31468ace51b6c/lib/sqlalchemy/engine/interfaces.py#L224-L229

File: doc/source/whatsnew/v2.1.0.rst [hunk 2]
```diff
@@ -219,6 +219,7 @@ MultiIndex
 
 I/O
 ^^^
+- Bug in type hint for ``params`` argument in :func:`read_sql_query` (:issue:`49716`)
```
[the remark this answers is not in the document]
Comment (Tolker-KU) [reply doc/source/whatsnew/v2.1.0.rst:222]: Got it. Will remove it :)

**Comment by Tolker-KU** [comment]
CI failed due to a GitHub issue. Can someone rerun the checks?

**Review by mroeschke — APPROVED** [review]

**Comment by mroeschke** [comment]
Thanks @Tolker-KU

**Pull request merged**